### ISSUE TITLE
Extremely slow compilation and high resource usage, even on v4 with turbopack

### ISSUE THREAD

**Issue by oprypkhantc** [shuding/nextra#3675]
**Describe the bug**

Hey. 

On the latest v4 release (version .22), the nextra-docs template takes >60sec to render the "Documentation" page, while using 300-400% CPU and 1.8GB of RAM on a modern MacBook Pro.

In of itself, a cold start taking some time isn't a huge problem; but even next renders (after changing a single page of content) still take >10sec to complete, all while using 2GB of RAM. Turbopack doesn't help at all here; it just uses 8x-10x times RAM, without any visible improvement to render time.

There is an old closed issue, mentioning that NextJS development is slow, but this really isn't the case with an example app from NextJS - where builds take <100ms. The difference is more than 100x in this case between a clean example NextJS app and a Nextra Docs template, which seems suspicious :)

**To Reproduce**

1. Clone the `examples/docs` template from `v4-v2` branch
2. Install deps, run the NextJS server: `yarn run next dev`
3. Navigate to `http://localhost:3000/docs`, wait for it to render
4. Change anything in `content/index.mdx`
5. See the change after 10-20sec

**Expected behavior**

Rerenders should probably be faster.

**Screenshots**

An example of how much resources initial render with Turbopack uses (some of it is shared memory, but still):

<img width="1129" alt="Screenshot 2024-11-08 at 23 59 44" src="https://github.com/user-attachments/assets/2936a0f3-1ec0-416b-83e7-4861577402a4">

**Comment by dimaMachina** [comment]
Changes from `content` folder may be slow compared to `page.{jsx,tsx,mdx}` convention because it uses dynamic import

https://github.com/shuding/nextra/blob/bf029ce8c2208a2c37c55dbbd67f35b56bc3ce98/packages/nextra/src/client/pages.ts#L11-L13

And it seems like Webpack/Turbopack precompiles all files from it.

Did you notice the same problems on migrated nextra.site website ([`docs` folder](https://github.com/shuding/nextra/tree/v4-v2/docs))?

> See the change after 10-20sec

On my mac I don't need to wait this time, everything happens less for a second

P.S. Also, solving [this issue](https://github.com/vercel/next.js/issues/71959) should improve performance for Turbopack projects

**Comment by oprypkhantc** [comment]
> Did you notice the same problems on migrated nextra.site website ([docs folder](https://github.com/shuding/nextra/tree/v4-v2/docs))?

No, in fact it's as fast as you describe it - hot reload updates take less than a second, and the compilation itself is ~100ms. It's as quick as I expect it to be :)

So having no previous experience with NextJS, as I understand it, the docs example has a fallback catch-all route that causes all of the assets to be recompiled, and the migrated Nextra site just uses Next's static routing? And the solution is to ditch the dynamic import and dynamic routing?

**Comment by dimaMachina** [comment]
> and the migrated Nextra site just uses Next's static routing?

Nextra 4 has 2 _modes_

- [Next.js App Router `page.{jsx,tsx}` convention](https://nextjs.org/docs/app/building-your-application/routing/pages) but also with `page.{md,mdx}` extensions
- `content` dir, you may think is as [Next.js Pages Router](https://nextjs.org/docs/pages/building-your-application/routing/pages-and-layouts) which rendered via single catch-all `[[...mdxPath]].{jsx,tsx}` route

> [!TIP]
> They can work separately and together.

---

> the docs example has a fallback catch-all route that causes all of the assets to be recompiled

I can't give the right response to this question...

When I add console logs in Nextra's Webpack loader, and make some changes in MDX files - I see logs with [the file path of the changed file](https://github.com/shuding/nextra/blob/bf029ce8c2208a2c37c55dbbd67f35b56bc3ce98/packages/nextra/src/server/loader.ts#L82) (only the initial loading of the project contains logs from all files because Nextra builds `pageMap` - array of existing files/folder and their exported front matter or `metadata`, to show proper sidebar and navigation)

> [!NOTE]
>
> `content` dir was made an alternative to [Next.js Pages Router](https://nextjs.org/docs/pages/building-your-application/routing/pages-and-layouts) to make smooth migration, it's still experimental so it may have some performance issues, maybe some folks in the future will help with the improvement of this mode

**Comment by oprypkhantc** [comment]
> Nextra 4 has 2 modes

Thank you, that cleared it up for me.

> When I add console logs in Nextra's Webpack loader, and make some changes in MDX files - I see logs with [the file path of the changed file](https://github.com/shuding/nextra/blob/bf029ce8c2208a2c37c55dbbd67f35b56bc3ce98/packages/nextra/src/server/loader.ts#L82) (only the initial loading of the project contains logs from all files because Nextra builds pageMap - array of existing files/folder and their exported front matter or metadata, to show proper sidebar and navigation)

I'll try to debug it later. However, there's one interesting finding: when I start the example `docs` app from the cloned Nextra repository, with `package.json` referencing `nextra*` dependencies from `workspace`, it flies. It's fast. Just as fast as the website app. 

However, if I extract the demo into a separate folder (not a subpath of the Nextra repository), then replace those `workspace` versions with `4.0.0-app-router.22` and `4.0.0-app-router.22`, and install dependencies with yarn - it suddenly is slow again.

Which makes me think the problem is either in the commits after version `4.0.0-app-router.22`, or in the dev/prod build 🤔

**Comment by dimaMachina** [comment]
Interesting, thank for investigation, will be appreciated if you could debug or provide reproduction, so I could debug too when I'll have time 🙏

**Comment by oprypkhantc** [comment]
Ha. I found the culprit. It's the `repository.getFileLatestModifiedDateAsync`. No wonder you couldn't reproduce it - our repo has >250k commits, so simply getting a modification date (for an uncommitted file) takes ~12sec. Commenting out the git stuff fixes the problem completely, and so does committing a file. 

Looking through the `simple-git`'s code, it looks like as long as the file that's being checked was committed recently, it will be fast, but I'm sure that it will become slow again once it has tens of thousands of commits to go through before finding the file.

Unfortunately I do not see a way to disable these checks. I can see how it can be useful, but not at the expense of performance 😄 I'm not much of a JS dev, but I can PR an option to disable the Git/last edit stuff - something like a `timestamps: false` option in the config?

**Comment by dimaMachina** [comment]
Amazing @oprypkhantc, I can reproduce it too!
<img width="636" alt="image" src="https://github.com/user-attachments/assets/91ad65bb-e5fc-47f4-8567-82678f062adf">

I think the simplest fix will be to cache the result of `repository.getFileLatestModifiedDateAsync`

**Comment by oprypkhantc** [comment]
Even with caching, the initial build would still take minutes instead of seconds, due to having to look up each individual file 🤔  Maybe only calculate timestamps for production builds?

**Comment by dimaMachina** [comment]
yeah, changed, try https://github.com/shuding/nextra/releases/tag/nextra%404.0.0-app-router.24

**Comment by dimaMachina** [comment]
@oprypkhantc Even though we haven't calculated the last edit time for dev, I still feel we need some static value because we have this field in `_meta` file to hive this `timestamp` https://nextra-v2-164w6708l-shud.vercel.app/docs/docs-theme/built-ins/layout#toggle-visibility-for-timestamp

So it will be better for dx to see something on dev, something like `Date.now()`

**Comment by oprypkhantc** [comment]
Indeed working as expected with .24 :) Thank you!

**Comment by oprypkhantc** [comment]
> So it will be better for dx to see something on dev, something like `Date.now()`

Yes, makes sense. Dev builds will be closer to production, which is always good

**Comment by oprypkhantc** [comment]
By the way, @dimaMachina, are you also seeing that high of a resource usage? At peaks running the dev server on the docs example uses >2GB, and even after peaks it stays at ~1GB while doing nothing. 

I know that Webpack is very inefficient,  and I even enabled Next's `webpackMemoryOptimizations` setting, but it still seems a bit high of a usage. Also, since Turbopack is written with Rust, it should be super efficient, but in practise it just makes things worse by using ~2.5GB while again doing nothing at all.

Also I understand this may not be Nextra's issue at all, but it's hard to tell with these kind of issues. Especially given that an example NextJS app (with TypeScript and Tailwind) only uses ~600MB with Turbopack. Anyway, I first wanted to confirm that you're seeing this too, and that the issue you mentioned (https://github.com/vercel/next.js/issues/71959) is not related. Then I'll report this to NextJS and see what they say :)

**Comment by dimaMachina** [comment]
Hm.. I don't know how I can help with this issue, can you try to test:
- resource usage of Nextra website with only 1 page
- comparison of resource usage with [`@next/mdx`](https://github.com/vercel/next.js/tree/canary/packages/next-mdx) for example

**Comment by oprypkhantc** [comment]
Nextra with just the `blog` page is still the same - 2GB peak / 1GB after without Turbopack, and 2.5GB with Turbopack.
NextJS with `@next/mdx` uses 800MB with Turbopack, and all of my CPUs :( 

<img width="1119" alt="image" src="https://github.com/user-attachments/assets/29938484-766e-417c-b479-5b563b8f90ec">

The CPU issue then is a 100% Turbopack's fault then; memory usage - can it be caused by additional dependencies of Nextra? I did include Tailwind in the example NextJS app, but other than that it's the clean example app

**Comment by dimaMachina** [comment]
Try to disable one by one remark/rehype plugins to see which causes an issue https://github.com/shuding/nextra/blob/v4-v2/packages/nextra/src/server/compile.ts#L225

**Comment by oprypkhantc** [comment]
Even removing literally all plugins (except for recma ones) and even moving to `.md`, memory usage still peaked at 2.25GB and stays at 1.9GB.

Some plugins I didn't remove because they set some properties on the AST - I just made sure those set static values and do nothing else.

<img width="951" alt="image" src="https://github.com/user-attachments/assets/c3918cb1-c983-4382-b2e8-6e2415aa4458">

**Comment by dimaMachina** [comment]
Thank you 🙏 no more ideas for now

**Comment by dimaMachina** [comment]
maybe it's due this

https://github.com/shuding/nextra/blob/c095f98c8bf09bba6c135cd3d8cf48debfb08570/packages/nextra/src/server/loader.ts#L124-L127

we need to call only 1 time outside of loader

**Comment by jamesone** [comment]
I am also seeing this issue using the following version:

```
"dependencies": {
    "@heroicons/react": "2.2.0",
    "next": "^15.1.7",
    "nextra": "^4.2.11",
    "nextra-theme-blog": "^4.2.11",
    "react": "18.2.0",
    "react-dom": "18.2.0"
  },
  "devDependencies": {
    "autoprefixer": "10.4.20",
    "pagefind": "^1.3.0",
    "@tailwindcss/postcss": "^4.0.6",
    "tailwindcss": "^4.0.6"
  }
```

When running `pnpm build`, I'll get the following stack failure:

```
 NODE_OPTIONS="--max-old-space-size=8192" next build

   ▲ Next.js 15.1.7
   - Experiments (use with caution):
     · turbo

   Creating an optimized production build ...

<--- Last few GCs --->

[81747:0x140008000]   133523 ms: Mark-Compact 8087.6 (8234.8) -> 8073.8 (8236.8) MB, pooled: 0 MB, 6074.08 / 0.00 ms  (average mu = 0.045, current mu = 0.001) allocation failure; scavenge might not succeed
[81747:0x140008000]   140000 ms: Mark-Compact 8089.6 (8236.8) -> 8075.8 (8238.8) MB, pooled: 0 MB, 6471.12 / 0.00 ms  (average mu = 0.023, current mu = 0.001) allocation failure; scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
----- Native stack trace -----

```

**Comment by chrisdburr** [comment]
Hi\! I'm experiencing slow compilation on Nextra 4.6.1 with Next.js 15.5.7 and did some debugging.

**Environment:**
- Nextra: 4.6.1
- Next.js: 15.5.7
- macOS (Apple Silicon)
- Repository: ~2,200 commits

**Debugging results:**

I tested with minimal content (1 simple MDX file, no imports) vs a normal Next.js route:

| Route | Bundler | Compile time | Modules |
|-------|---------|--------------|---------|
| `/docs` (catch-all, 1 MDX) | Turbopack | 36.6s | - |
| `/docs` (catch-all, 1 MDX) | Webpack | 53.2s | 13,284 |
| `/docs` (catch-all, full) | Turbopack | 53-70s | - |
| `/discover` (normal route) | Turbopack | 2s | - |

```
○ Compiling /docs/[[...slug]] ...
✓ Compiled /docs/[[...slug]] in 53.2s (13284 modules)
```

Even with just 1 simple MDX file, the catch-all route compiles 13,284 modules and takes 36+ seconds. A comparable normal Next.js route compiles in ~2 seconds.

Since I only have ~2,200 commits, the git timestamp fix from `.24` shouldn't apply here. Is there any progress on optimising the `content/` directory mode compilation?

**Comment by dacevedo12** [comment]
I'm seeing increased memory consumption by just pressing ctrl S on a mdx file over and over (without changes), 5GB+ and never stops

**Comment by 03-CiprianoG** [comment]
Sharing measured data from a large production `content/`-mode site, since the recent reports here — especially the "ctrl-S → 5 GB and never stops" one — match exactly what we hit, and we landed on a dev-memory mitigation that isn't mentioned anywhere in this thread.

### Context

This is the internal docs for a fairly large SaaS platform: **59 MDX files, ~1,190 fenced code blocks (216 of them Mermaid diagrams), ~2.1 MB of content**, all served through the single `content/` catch-all `[[...mdxPath]]` route. So it's a heavy mix of syntax-highlighted code and client-rendered diagrams — pretty much the worst case for this route.

- **Env:** Next 16.2.4, Nextra 4.6.1, Node 24, Turbopack dev, WSL2 (22 GB).

### Symptom

Dev `next-server` grew **~600 MB per visited page and never evicted**. Browsing ~4 heavy pages took it from ~3.6 GB to OOM-crash. Repeatedly saving a single file climbed the same way without ever coming back down — identical to what @dacevedo12 described.

### Our debugging journey (what it was *not*)

We measured per-process RSS (`VmRSS`/`VmHWM`) under a scripted curl drive across the heavy routes, and ruled out the usual suspects one by one:

- **Not Shiki alone** — gating `codeHighlight` to prod (`codeHighlight: process.env.NODE_ENV === 'production'`) helps per-page compile cost a lot, but on its own it did **not** stop the per-route accumulation.
- **Not our content/config being "wrong"** — we diffed against a sibling Nextra 4.6.1 site with the *same* route/`mdx-components`/layout wiring. It grows the same way, just from a smaller base (it topped out ~6 GB over the same 8-route drive instead of OOMing — purely because it has ¼ the content).
- **Not the git-timestamp path** (the `.24` fix) — already on 4.6.1.
- **Not a leak** — memory is fully reclaimed on process exit; it's a one-way climb of retained, never-evicted compiled routes.

So the driver is exactly what's described above: the catch-all pulls the content graph in, and Turbopack dev doesn't evict it.

### What actually bounded it (the part not yet mentioned here)

Two Next experimental flags:

```js
// next.config.mjs
export default withNextra({
  experimental: {
    preloadEntriesOnStart: false,   // <-- the decisive one
    turbopackSourceMaps: false,     // trims per-module overhead in dev
  },
})
```

`preloadEntriesOnStart: false` ([Next docs](https://nextjs.org/docs/app/guides/memory-usage#preloading-entries)) stops Next from eagerly preloading every page's modules into memory at server start, which keeps dev compilation genuinely on-demand. Same 8-route stress run, same content, before vs after:

| | before | after |
|---|---:|---:|
| `next-server` peak | **~8.4 GB** (OOM around route 4) | **~2.0 GB** |
| 28 distinct heavy routes back-to-back | crashes | **~2.3 GB, basically flat** |

Per-route growth dropped from ~600 MB/page to ~30–80 MB/page. It turned a hard OOM into a stable ~2 GB dev session.

### Controlled A/B: `content/` catch-all vs `page.{mdx}` static mode

Because this thread raises whether migrating off `content/` mode fixes it, we tested it directly. We rebuilt our heaviest section (11 files / 290 code blocks) as static `app/<route>/page.mdx`, identical content and config, and measured the same way:

| Mode | baseline (1 route) | after all routes | HMR recompile (edit 1 page) |
|---|---:|---:|---:|
| `content/` catch-all | ~1.7 GB | ~2.0 GB | multi-second |
| `page.mdx` static | ~1.3 GB | ~2.3 GB | **~0.2 s** |

The **~2 GB memory floor is essentially mode-independent** — switching to page-file routing did *not* lower it. Where page-file mode wins is **compile/HMR**: a single named route compiles instead of the catch-all graph (consistent with @chrisdburr's "13,284 modules for 1 MDX file" finding), so edits hot-reload in ~0.2 s instead of seconds.

### Honest scope

These flags **bound dev memory** — they turn the OOM/non-eviction into a stable ~2 GB — but they do **not** fix the underlying `content/`-mode *compile cost*. @chrisdburr's 13k-modules / 50s+ compile time is unchanged by them; that still looks like the real thing to fix in `content/` mode. Posting this in case it unblocks others hitting the OOM in the meantime, and because the `preloadEntriesOnStart` lever doesn't seem to be known here.
